### ISSUE TITLE
[TraceQL] Support attribute existence filter

### ISSUE THREAD

**Issue by joe-elliott** [grafana/tempo#2188]
**Is your feature request related to a problem? Please describe.**
There should be a way to query for the existence of an attribute without asserting a value. Perhaps something like:

```
{ span.att = nil }
{ exists(span.att) }
{ span.att }
```

Note that the last one overloads querying a boolean attribute for true or false values. If we want to go that route we should think through it carefully.

**Comment by lanapouney** [comment]
Hi, I am looking for a way to query that an attribute does not exist for a span, could this be achieved via { !exists(span.att) } or would you propose another method for this?

**Comment by joe-elliott** [comment]
Unfortunately this does not currently exist. I want it too, but haven't found the time to add it. Do you have a preferred syntax above?

If you know the type of your attribute you can do something like `{ span.att != ""}` or  `{ span.att != 0 }'. I know that's not great, but it's the only current work around.

**Comment by cyrille-leclerc** [comment]
I prefer `{ span.att = nil }` and `{ span.att = nil }` because they reuse the existing TraceQL [comparison operators](https://grafana.com/docs/tempo/latest/traceql/#comparison-operators) and work well IMO for auto-completion

**Comment by cedricziel** [comment]
I'm putting my money on `{ .attr = nil }` as well because it's simple and intuitive

**Comment by adirmatzkin** [comment]
Just making sure I don't miss out on something here...
By querying for a value of "nil" for some attribute `x` (syntax as `{ span.x = nil }`), we would assume `x` doesn't even exist?
How would we differentiate between the 2 cases:
1. x is not even a field in a span
2. x **is** a field in a span, but **its value is** None/Null/Nil.

If we're about to use "nil", "none", or "null" for different meanings here - let me raise a red flag, because it would be confusing and not intuitive for users...

Unless I'm missing something - I feel like the `{ exists(span.x) }` syntax would be better here due to the need to differentiate between the above 2 cases.

**Comment by cyrille-leclerc** [comment]
Great verification.

From what I understand, null value is not allowed for hotel attributes and
it means to suppress the attribute.
I have to read in depth
https://github.com/open-telemetry/opentelemetry-specification/issues/797

**Comment by cyrille-leclerc** [comment]
Confirmed, otel attributes MUST be non null.

See
https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/common

"Attribute values of null are not valid and attempting to set a null value
is undefined behavior."

**Comment by adirmatzkin** [comment]
Nice, so I did miss something 🙃 , thanks @cyrille-leclerc!

If that's the case, and it looks like the first two options Joe mentioned in the original comment are the leading ones, I'd like to help push this feature forward - which means at this point helping out getting to a decision 😃.

Here are (IMO) the main considerations we should focus on and take into account when dealing with this kind of syntax question:

**1. Expressiveness:** Allow users to express a wide range of queries and operations in different query patterns.
**2. Readability, simplicity, and consistency:** Keep it easy and simple to learn/use/read-understand/maintain, by leveraging familiar conventions that will make it intuitive.
**3. Performance:** Although syntax won't directly impact performance, design decisions can affect execution efficiency by enabling/limiting optimizations in query processing or execution.

The more I think about it - I'm getting convinced the `{ span.att = nil }` syntax will be the better choice.
- It is more consistent with the current syntax - also saving us unnecessary complications and introducing new functions and overall ways to express queries.
- As for expressiveness - I see this feature as a very specific one. Basically allowing 2 (that are kind of the same) querying abilities - `{ span.attr = nil }` for "not exists", and `{ span.attr != nil }` for "exists".
- As for performance - I'm not (yet) familiar with TraceQL engine, perhaps @joe-elliott could help out with this consideration.

I think that unless the performance parameter is a win for the `exists(span.attr)` - there is a winning option and we can continue moving! 💪🏼
Looking forward to hearing an opinion from someone whose familiar with the TraceQL engine 🙃

**Comment by cyrille-leclerc** [comment]
Adding `!= nil` would bring so much value that the inconsistency of lacking of `= nil` is IM£O a minor disadvantage.

**Comment by mdisibio** [comment]
> = nil will be quite difficult. The fetch layer all the way down to Parquet is entirely built around finding row numbers that match conditions. To make this work we'd need to pull a list of all row numbers at the span level, and then all row numbers where != nil is true and negate them.

Agree, can't think of a more efficient algorithm and that type of query would always be much slower than `!= nil`.  But I do think it's achievable with the current storage layer.  Fetch something like status (all the row numbers), the attr with no filtering (opnone), and then let the engine determine the matches.

**Comment by knylander-grafana** [comment]
The output from the work on this issue can be documented with https://github.com/grafana/tempo/issues/2188

**Comment by mdisibio** [comment]
> "find all `span` attributes, and in case span has `attr`, it must be `"test"`"

Something like this: `{ name = "span" && (span.attr = nil || span.attr = "test") }` ?  That would work as expected when support for `= nil` is added.

**Comment by zohnannor** [comment]
That's unfortunate. The lack of ability to find `nil` values cannot be solved with a workaround and thus requires me to use two different TraceQL queries.

**Comment by mdisibio** [comment]
Hi everyone, I discovered a way to achieve this using the existing language and wanted to share it.   I think this could be useful if anyone has a pressing need for the functionality today, it's not meant to be a replacement for the better options discussed above.

The query is:  `{} | select(.attr) | { !(.attr != nil)}`       

To explain:  We can already pull back an attribute optionally using select().  The span will be populated with the value if it exists.  The trick is that those results can be run through another filter to get just the spans that were missing it.    The second clause acts like `.attr = nil` but that specifically can't be used yet since our query validation catches it.

Here's a screenshot using https://github.com/grafana/xk6-client-tracing showing that it works.  All of the results are for `cache` or `identityService` which do not populate these values (and you can see on the right-hand side http.url is missing):
![Image](https://github.com/user-attachments/assets/ace3f754-9d97-4b2c-994c-b0c6b55c925a)

**Comment by 09jvilla** [comment]
@mdisibio , I noticed that 

`{} | select(.attr) | { !(.attr != nil)}` worked when I tried it, but that `{} | select(resource.attr) | { !(resource.attr != nil)}` and `{} | select(span.attr) | { !(span.attr != nil)}` did not. Essentially, specifying a scope in front of the attribute results in the query executing but returning no matches. Is this expected? I would think that if a span is missing an attribute entirely, then it also is definitely missing that attribute when you look at the resource scope, or the span scope, or any other scope in fact.

**Comment by mdisibio** [comment]
> specifying a scope in front of the attribute results in the query executing but returning no matches

Good catch, that's a bug.  Specifying a scope shouldn't matter.   Not 100% unexpected since this was kind of a hacky work-around already.  Checking briefly, it has to do with the optimizer trying things when a scope is specified, that it's not able to do with unscoped (.foo).

**Comment by diegocejasprieto** [comment]
correct me if I'm wrong but wouldn't it be easier something like this?

`{} | { (.attr != nil) }` <-- for searching for an EXISTING attribute
`{} | { !(.attr!= nil) }` <-- for searching for a NON EXISTING attribute.

I'm testing this two queries for an Opentelemetry stack where I need to filter out for a non existing resource attribute and the second option seems to work for me

**Comment by mdisibio** [comment]
@diegocejasprieto I tested the second query and it does not work for me.  `{} | { !(resource.foo != nil) }` returns 0 results.  But the version listed above `{} | select(resource.foo) | { !(resource.foo != nil) }` returns everything.   The `select(...)` is required to make the attribute lookup optional (internal details of how traceql works).  This work-around is only needed until we add official support for `= nil`. 

The first query there is already official support to check for existence with nice syntax `{ resource.foo != nil }`.

**Comment by Jxlam-pdx** [comment]
@ie-pham just synced with @cyrille-leclerc and the use case we want to prioritize first is `{ resource.attr = nil }`.  It came up earlier at our Weekly that we may want to separate out the remainder of the work into a few issues.  Let prioritize this one and we can put work required for intrinsic into a different ticket.

**Comment by cyrille-leclerc** [comment]
To better support popular OTel use cases, we would like to support scenarios where some teams would NOT define resource attributes `deployment.environment.name` or `service.namespace`.

For example filter all traces matching

**Undefined `service.namespace`**

```
resource.service.name="frontend" && resource.service.namespace=nil
````
versus
```
resource.service.name="frontend" && resource.service.namespace="webshop"
````

**Undefined `deployment.environment.name`**

```
resource.service.name="frontend" && resource.deployment.environment.name=nil
````
versus
```
resource.service.name="frontend" && resource.deployment.environment.name="production"
````

**Overall experience**

Beyond these two "unit tests" above, the goal is to have a Grafana dashboard with 3 variables `deployment.environment.name`, `service.namespace`, `service.name` displaying matching traces in a panel supporting undefined `deployment.environment.name` or `service.namespace`.

An example of such a dashboard can be found with the Grafana community dashboard [Lightweight APM for OpenTelemetry](https://grafana.com/grafana/dashboards/22784-lightweight-apm-for-opentelemetry/).

**Comment by Jxlam-pdx** [comment]
Grafana Variables are built around Mimir/Loki, this ask is not easily accomplished with Tempo. There is a limitation on how/where we put the " ". Even if we added != nil, the use case above still won't work. @cyrille-leclerc let's setup some time to brainstorm.

**Comment by cyrille-leclerc** [comment]
Thanks @Jxlam-pdx. 
I'm totally with you that TraceQL implementing `my_attr = nil` is not enough to solve the problem of Grafana dashboards variables supporting "All", "multi-select", and "not defined" across languages beyond PromQL and LogQL.

I did a quick review of the query languages of a few popular Grafana datasource on:
* https://github.com/grafana/grafana/issues/102236

I'm working with @thanos-karachalios and @dprokop on this. Your help would be greatly appreciated.
